linked_list/linked_list.py

```diff
@@ -1,75 +1,170 @@
 
 # Defines a node in the singly linked list
+from tkinter.messagebox import NO
+
+
 class Node:
 
-    def __init__(self, value, next_node = None):
+    def __init__(self, value, next_node=None, previous_node=None):
         self.value = value
         self.next = next_node
+        self.previous = previous_node
 
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None # keep the head private. Not accessible outside this class
+        self.tail = None
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if self.head is None:
+            return None
+        else:
+            return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        if self.head is None:
+            self.head = self.tail = Node(value)
+        else:
+            new_node = Node(value, self.head)
+            self.head.previous = new_node
+            self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if self.head is None:
+            return False
+
+        # # Traverse the list until you find the node at index -1
+        current = self.head
+        while current:
+            if current.next is None:
+                if current.value == value:
+                    return True
+                else:
+                    return False
+            elif current.value == value:
+                return True
+            else:
+                current = current.next
+        return False
+    
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):

```

Comment:
👍 Just a note, you could also maintain a `length` attribute to keep track of how many nodes the structure has bringing this down to O(1) time complexity for a bit of additional coding.